### PR TITLE
Pin docker-cli version to the 17.06-ce release version

### PR DESCRIPTION
**Pull request by dnephin**

### CODE REVIEW

**Comment by dnephin** [comment]
Thanks @johnstep for the help with the windows version in `hack/make.ps1`!

File: hack/make.ps1 [hunk 1]
```diff
@@ -398,40 +398,11 @@ Try {
...
-            Finally {
-                # Always restore GOPATH and remove the temporary directory.
-                $env:GOPATH = $saveGOPATH
-                Remove-Item -Force -Recurse $tempLocation
-            }
+            Invoke-WebRequest -Uri https://download.docker.com/win/static/edge/x86_64/docker-17.06.0-ce.zip -OutFile docker.zip
+            Expand-Archive docker.zip
+            Remove-Item docker.zip
+            Move-Item docker\docker\docker.exe $root\bundles\docker.exe
+            Remove-Item -Recurse docker
```
Comment (johnstep) [inline hack/make.ps1]: Replace with the following, which supports the environment variables and handles errors:

```powershell
            # Get the Docker channel and version from the environment, or use the defaults.
            if (-not ($channel = $env:DOCKERCLI_CHANNEL)) { $channel = "edge" }
            if (-not ($version = $env:DOCKERCLI_VERSION)) { $version = "17.06.0-ce" }

            # Download the zip file and extract the client executable.
            Write-Host "INFO: Downloading docker/cli version $version from $channel..."
            $url = "https://download.docker.com/win/static/$channel/x86_64/docker-$version.zip"
            Invoke-WebRequest $url -OutFile "docker.zip"
            Try {
                Add-Type -AssemblyName System.IO.Compression.FileSystem
                $zip = [System.IO.Compression.ZipFile]::OpenRead("$PWD\docker.zip")
                if (-not ($entry = $zip.Entries | Where-Object { $_.Name -eq "docker.exe" })) {
                    Throw "Cannot find docker.exe in $url"
                }
                [System.IO.Compression.ZipFileExtensions]::ExtractToFile($entry, "$PWD\bundles\docker.exe", $true)
            }
            Catch [Exception] {
                Throw $_
            }
            Finally {
                # Always close and delete the zip file.
                if ($zip) { $zip.Dispose() }
                Remove-Item -Force "docker.zip"
            }
```

**Review by johnstep — CHANGES_REQUESTED** [review]

**Comment by jhowardmsft** [comment]
I'm not sure I understand why this is being pinned? This way, and without a `make.ps1` equivalent in the CLI binary (which we really should have), it means there's no easy way on Windows to build the CLI anymore at an arbitrary commit to match daemon changes. IMO I don't think that's really what we need and this is a bit of a backwards step.

**Comment by dnephin** [comment]
> I'm not sure I understand why this is being pinned? 

It has always been pinned, that is not changing. The change is to use a release version instead of some arbitrary git sha.

This has been planned for a while as part of the moby and docker split. Moby should not depend on docker. Unfortunately most of the integration test suite depends on a client binary, so until we've completed the transition to API tests we need to keep a client binary around. By pinning a specific version of the binary we prevent tests from failing due to unrelated changes. The previous pin to git sha was never meant to support building cli binaries. We were just waiting until we had a release binary we could use.

> without a make.ps1 equivalent in the CLI binary  [repository] there's no easy way on Windows to build the CLI anymore at an arbitrary commit to match daemon changes. 

This is something we should fix in the `docker/cli` repo.  https://github.com/docker/cli/issues/183 is a related issue, which should fix this problem as well . These concerns are outside the scope of the `moby/moby` repo. The `Dockerfile.windows` in this repo shouldn't be used for building the cli. It's completely coincidental that it happens to work right now.

I can look at implementing https://github.com/docker/cli/issues/183 next week.

**Review by johnstep — APPROVED** [review]
The Windows changes look good. I agree with @jhowardmsft that we need a native Windows way to build the client (docker/cli), and will look into that soon.

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

File: hack/dockerfile/install-binaries.sh [hunk 2]
```diff
@@ -47,11 +47,15 @@ install_proxy() {
...
-	git clone "$DOCKERCLI_REPO" "$GOPATH/src/github.com/docker/cli"
-	cd "$GOPATH/src/github.com/docker/cli"
-	git checkout -q "$DOCKERCLI_COMMIT"
-	go build -o /usr/local/bin/docker github.com/docker/cli/cmd/docker
+	DOCKERCLI_CHANNEL=${DOCKERCLI_CHANNEL:-edge}
+	DOCKERCLI_VERSION=${DOCKERCLI_VERSION:-17.06.0-ce}
+	echo "Install docker/cli version $DOCKERCLI_VERSION from $DOCKERCLI_CHANNEL"
+
+	url=https://download.docker.com/linux/static
+	curl -Ls $url/$DOCKERCLI_CHANNEL/x86_64/docker-$DOCKERCLI_VERSION.tgz | \
```
Comment (tophj-ibm) [inline hack/dockerfile/install-binaries.sh]: p/z failures are real. While s390x and armhf are in that channel and can be referenced here, ppc64le, arm64, and to a lesser extent i386 aren't so they would still have to be built from src.

If the cli binary corresponds directly to the upstream repo, it might make sense to have version tags in the cli repo.

Comment (dnephin) [inline hack/dockerfile/install-binaries.sh]: The release branches and version tags are in https://github.com/docker/docker-ce. Since the release branch receives cherry-picked patches there is unfortunately no corresponding commit in `docker/cli`.

Comment (dnephin) [inline hack/dockerfile/install-binaries.sh]: I guess for powerpc we will need to build from source

File: hack/dockerfile/install-binaries.sh [hunk 3]
```diff
@@ -52,14 +52,29 @@ install_dockercli() {
 	echo "Install docker/cli version $DOCKERCLI_VERSION from $DOCKERCLI_CHANNEL"
 
 	arch=$(uname -m)
-	echo "Docker CLI ARCH $arch"
+	# No official release of these platforms
+	if [[ "$arch" == "ppc64le" ]]; then
```
Comment (tophj-ibm) [inline hack/dockerfile/install-binaries.sh]: Can you do if not eq x86_64/armhf/s390x here instead? That should cover all the arm variants.

**Comment by dnephin** [comment]
Ok, updated to build on any platform without an official release

File: hack/dockerfile/install-binaries.sh [hunk 4]
```diff
@@ -47,10 +47,31 @@ install_proxy() {
...
-	git clone "$DOCKERCLI_REPO" "$GOPATH/src/github.com/docker/cli"
-	cd "$GOPATH/src/github.com/docker/cli"
-	git checkout -q "$DOCKERCLI_COMMIT"
+	DOCKERCLI_CHANNEL=${DOCKERCLI_CHANNEL:-edge}
+	DOCKERCLI_VERSION=${DOCKERCLI_VERSION:-17.06.0-ce}
+	echo "Install docker/cli version $DOCKERCLI_VERSION from $DOCKERCLI_CHANNEL"
+
+	arch=$(uname -m)
+	# No official release of these platforms
+	if [[ "$arch" != "x86_64" ]] && [[ "$arch" != "s390x" ]] && [[ "$arch" != "armhf" ]]; then
```
Comment (tianon) [inline hack/dockerfile/install-binaries.sh]: The output of `uname -m` will likely never be `armhf` -- it'll be something like `armv5tel`, `armv6l`, `armv7l`, `armv8l`, `aarch64`, etc.  Using `go env GOARCH` or `go env GOHOSTARCH` will likely be more reliable.

Comment (dnephin) [inline hack/dockerfile/install-binaries.sh]: fun... That gives me `amd64`, which is not what we need either.

I'm just going to remove `armhf` from this list.

**Comment by dnephin** [comment]
this is green again

File: hack/dockerfile/install-binaries.sh [hunk 5]
```diff
@@ -47,10 +47,31 @@ install_proxy() {
 }
 
 install_dockercli() {
-	echo "Install docker/cli version $DOCKERCLI_COMMIT"
-	git clone "$DOCKERCLI_REPO" "$GOPATH/src/github.com/docker/cli"
-	cd "$GOPATH/src/github.com/docker/cli"
-	git checkout -q "$DOCKERCLI_COMMIT"
+	DOCKERCLI_CHANNEL=${DOCKERCLI_CHANNEL:-edge}
+	DOCKERCLI_VERSION=${DOCKERCLI_VERSION:-17.06.0-ce}
```
Comment (tophj-ibm) [inline hack/dockerfile/install-binaries.sh]: sgtm other than this should be 17.06.1-ce now 😭

Comment (dnephin) [reply hack/dockerfile/install-binaries.sh]: I don't think so!

We want to freeze this at 17.06, we should never update this value.

**Comment by dnephin** [comment]
@vdemeester @johnstep still looks good to you?

File: integration-cli/docker_cli_service_logs_test.go [hunk 6]
```diff
@@ -40,7 +40,7 @@ func (s *DockerSwarmSuite) TestServiceLogs(c *check.C) {
 	// make sure task has been deployed.
 	waitAndAssert(c, defaultReconciliationTimeout,
 		d.CheckRunningTaskImages, checker.DeepEquals,
-		map[string]int{"busybox": len(services)})
+		map[string]int{"busybox:latest": len(services)})
```
Comment (tiborvass) [inline integration-cli/docker_cli_service_logs_test.go]: @dnephin why these changes? I wonder if you take 17.06.1-ce then you may not need to change this.

Comment (dnephin) [reply integration-cli/docker_cli_service_logs_test.go]: Why do you think this would have reverted in 17.06.1?

**Review by vdemeester — APPROVED** [review]
still LGTM 🐮

File: integration-cli/docker_cli_service_logs_test.go (hunk 6 above)
Comment (dnephin) [reply integration-cli/docker_cli_service_logs_test.go]: @tiborvass as you can see from the below output, this is the new behaviour. There's nothing to fix, and it's any different in new releases:

```
$ docker-17.05 service create --name test1705  --label tag-test alpine
tet0jh8j5ygy7me632d3ddrvx

$ docker-17.06 service create --name test1706  --label tag-test alpine
s4cvxf4za8q74nfx5taobs9yk

$ docker-17.06.1 service create --name test17061  --label tag-test alpine
oyyc1i71kn7dusk012frx6ouz

$ docker-17.07 service create --name test1707  --label tag-test alpine
9d0qo4whxcr3fojdk4xwpsd9i

$ docker service ls -f label=tag-test
ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
9d0qo4whxcr3        test1707            replicated          0/1                 alpine:latest
oyyc1i71kn7d        test17061           replicated          0/1                 alpine:latest
s4cvxf4za8q7        test1706            replicated          0/1                 alpine:latest
tet0jh8j5ygy        test1705            replicated          0/1                 alpine
```


```

Comment (tiborvass) [reply integration-cli/docker_cli_service_logs_test.go]: @dnephin I'm concerned about the tests we run for releases cc @andrewhsu @seemethere will this change impact tests on docker-ce ? If so, we need a good solution for the problem before merging.

Comment (seemethere) [reply integration-cli/docker_cli_service_logs_test.go]: @tiborvass @dnephin Kicked off a build to test integration tests with this PR: https://jenkins.dockerproject.org/job/docker-integration-tests/94/

Comment (dnephin) [reply integration-cli/docker_cli_service_logs_test.go]: @tiborvass @seemethere we run the release tests using a release binary, so the behaviour will match the updated test code.

This will actually fix two of the reported problems with that suite. See #34623 `TestServiceLogs` and `TestSwarmNetworkPluginV2` are fixed by this change.

**Review by seemethere — APPROVED** [review]
LGTM

**Review by vieux — APPROVED** [review]
LGTM

**Comment by Battleborn69** [comment]
Thank you for noticing

**Comment by Battleborn69** [comment]
I Agree

**Comment by seemethere** [comment]
@dnephin Actually upon closer inspection of this PR, after running integration tests, it seems that the `busybox:latest` is causing failures for `TestInspect` in `integration/service/inspect_test.go`

Particularly with these lines:

https://github.com/moby/moby/blob/e23965d620e1655f44d47edd8b5e08c2c384fd03/integration/service/inspect_test.go#L20-L21

Which outputs:
```
00:04:59.890 Running /go/src/github.com/docker/docker/integration/service
00:04:59.890 INFO: Testing against a local daemon
00:04:59.890 === RUN   TestInspect
00:05:00.454 --- FAIL: TestInspect (0.51s)
00:05:00.454 	daemon.go:285: [d29b9efd12c54] waiting for daemon to start
00:05:00.454 	daemon.go:317: [d29b9efd12c54] daemon started
00:05:00.454 	assertions.go:226: 
                          
	Error Trace:	daemon.go:503
00:05:00.454 		
	            	
			daemon.go:333
00:05:00.454 		
	            	
			inspect_test.go:123
00:05:00.454 		
	            	
			inspect_test.go:21
00:05:00.454 		
	Error:      	Received unexpected error:
00:05:00.454 		
	            	Error response from daemon: reference does not exist
00:05:00.454 		
	Messages:   	failed to download busybox
00:05:00.454 FAIL
```

Also, while it may be unrelated to this particular PR we are also seeing more failures when comparing test runs from yesterday to today.

<details> 
  <summary>Yesterday (August 29, 2017) test failures</summary>

```
$ curl -fsSL https://jenkins.dockerproject.org/job/docker-integration-tests/92/consoleText | grep "FAIL:" | sort -r | grep -v ' --- '
[DockerSwarmSuite] FAIL: docker_cli_swarm_unix_test.go:56: DockerSwarmSuite.TestSwarmNetworkPluginV2
[DockerSwarmSuite] FAIL: docker_cli_service_logs_test.go:24: DockerSwarmSuite.TestServiceLogs
[DockerSuite] FAIL: docker_cli_update_unix_test.go:286: DockerSuite.TestUpdateWithNanoCPUs
[DockerSuite] FAIL: docker_cli_run_unix_test.go:207: DockerSuite.TestRunAttachDetachFromInvalidFlag
[DockerSuite] FAIL: docker_cli_run_unix_test.go:1558: DockerSuite.TestRunWithNanoCPUs
[DockerSuite] FAIL: docker_cli_rmi_test.go:227: DockerSuite.TestRmiContainerImageNotFound
[DockerSuite] FAIL: docker_cli_rmi_test.go:144: DockerSuite.TestRmiImageIDForceWithRunningContainersAndMultipleTags
[DockerSuite] FAIL: docker_cli_rename_test.go:72: DockerSuite.TestRenameInvalidName
```

</details>

<details>
  <summary>Today's (August 30, 2017) test failures</summary>

```
$ curl -fsSL https://jenkins.dockerproject.org/job/docker-integration-tests/100/consoleText | grep "FAIL:" | sort -r | grep -v ' --- '
[DockerTrustSuite] FAIL: docker_cli_run_test.go:3190: DockerTrustSuite.TestTrustedRunFromBadTrustServer
[DockerTrustSuite] FAIL: docker_cli_run_test.go:3161: DockerTrustSuite.TestTrustedRun
[DockerTrustSuite] FAIL: docker_cli_create_test.go:329: DockerTrustSuite.TestTrustedCreateFromBadTrustServer
[DockerTrustSuite] FAIL: docker_cli_create_test.go:319: DockerTrustSuite.TestTrustedIsolatedCreate
[DockerTrustSuite] FAIL: docker_cli_create_test.go:293: DockerTrustSuite.TestTrustedCreate
[DockerSwarmSuite] FAIL: docker_api_swarm_test.go:349: DockerSwarmSuite.TestAPISwarmRaftQuorum
[DockerSuite] FAIL: docker_cli_update_unix_test.go:286: DockerSuite.TestUpdateWithNanoCPUs
[DockerSuite] FAIL: docker_cli_run_unix_test.go:207: DockerSuite.TestRunAttachDetachFromInvalidFlag
[DockerSuite] FAIL: docker_cli_run_unix_test.go:1558: DockerSuite.TestRunWithNanoCPUs
[DockerSuite] FAIL: docker_cli_rmi_test.go:227: DockerSuite.TestRmiContainerImageNotFound
[DockerSuite] FAIL: docker_cli_rmi_test.go:144: DockerSuite.TestRmiImageIDForceWithRunningContainersAndMultipleTags
[DockerSuite] FAIL: docker_cli_rename_test.go:72: DockerSuite.TestRenameInvalidName
[DockerRegistrySuite] FAIL: docker_cli_pull_local_test.go:451: DockerRegistrySuite.TestRunImplicitPullWithNoTag
[DockerRegistrySuite] FAIL: docker_cli_by_digest_test.go:142: DockerRegistrySuite.TestRunByDigest
[DockerRegistrySuite] FAIL: docker_cli_by_digest_test.go:129: DockerRegistrySuite.TestCreateByDigest
[DockerNetworkSuite] FAIL: docker_cli_network_unix_test.go:1812: DockerNetworkSuite.TestConntrackFlowsLeak
[DockerDaemonSuite] FAIL: docker_cli_daemon_plugins_test.go:190: DockerDaemonSuite.TestVolumePlugin
```

</details>

<details>
  <summary>Diff between the two</summary>

```
❯ diff <(curl -fsSL https://jenkins.dockerproject.org/job/docker-integration-tests/92/consoleText | grep "FAIL:" | sort -r | grep -v ' --- ') <(curl -fsSL https://jenkins.dockerproject.org/job/docker-integration-tests/100/consoleText | grep "FAIL:" | sort -r | grep -v ' --- ')
1,2c1,6
< [DockerSwarmSuite] FAIL: docker_cli_swarm_unix_test.go:56: DockerSwarmSuite.TestSwarmNetworkPluginV2
< [DockerSwarmSuite] FAIL: docker_cli_service_logs_test.go:24: DockerSwarmSuite.TestServiceLogs
---
> [DockerTrustSuite] FAIL: docker_cli_run_test.go:3190: DockerTrustSuite.TestTrustedRunFromBadTrustServer
> [DockerTrustSuite] FAIL: docker_cli_run_test.go:3161: DockerTrustSuite.TestTrustedRun
> [DockerTrustSuite] FAIL: docker_cli_create_test.go:329: DockerTrustSuite.TestTrustedCreateFromBadTrustServer
> [DockerTrustSuite] FAIL: docker_cli_create_test.go:319: DockerTrustSuite.TestTrustedIsolatedCreate
> [DockerTrustSuite] FAIL: docker_cli_create_test.go:293: DockerTrustSuite.TestTrustedCreate
> [DockerSwarmSuite] FAIL: docker_api_swarm_test.go:349: DockerSwarmSuite.TestAPISwarmRaftQuorum
8a13,17
> [DockerRegistrySuite] FAIL: docker_cli_pull_local_test.go:451: DockerRegistrySuite.TestRunImplicitPullWithNoTag
> [DockerRegistrySuite] FAIL: docker_cli_by_digest_test.go:142: DockerRegistrySuite.TestRunByDigest
> [DockerRegistrySuite] FAIL: docker_cli_by_digest_test.go:129: DockerRegistrySuite.TestCreateByDigest
> [DockerNetworkSuite] FAIL: docker_cli_network_unix_test.go:1812: DockerNetworkSuite.TestConntrackFlowsLeak
> [DockerDaemonSuite] FAIL: docker_cli_daemon_plugins_test.go:190: DockerDaemonSuite.TestVolumePlugin
```

</details>

**Comment by dnephin** [comment]
Thanks unrelated, the fix is here: #34682
